레스토랑 조회 구현하기

2scent opened the pull request:
안녕하세요. 이번 주도 잘 부탁드립니다.

우선 기본적인 리덕스 관련 코드 작성한 후에, 비동기 작업을 처리하기 위해 `api.js`와 `async-actions.js`라는 파일을 만들어서 각각 테스트 코드도 작성해봤는데 이런 식으로 하면 될까요? api에 대한 테스트는 강의에서 없었던 부분이라 약간 과한 것 같기도 하고 잘 모르겠네요.. 🤔

2scent committed:
Add redux settings

2scent committed:
Add initial state to reducer

2scent committed:
Add setRegions action

2scent committed:
Add setCatogories action

2scent committed:
Add setRestaurants action

2scent committed:
Add selectRegion action

2scent committed:
Add selectCategory action

2scent committed:
Refactor reducer to use actionHandlers

2scent committed:
Add action creators

action creator 추가하고, reducer test에서 사용

2scent committed:
Add api service

2scent committed:
Refactor api service

2scent committed:
Add async actions

passwd10 commented on src/services/api.js:
```diff
@@ -0,0 +1,11 @@
+const fetchApi = async (api) => {
+  const response = await fetch(`https://eatgo-customer-api.ahastudio.com${api}`);
+  const data = await response.json();
+  return data;
+};
```
함수로 분리해 중복을 최소화화셨네요 👍

passwd10 commented on src/services/api.test.js:
```diff
@@ -0,0 +1,76 @@
+
+import { fetchCategories, fetchRegions, fetchRestaurants } from './api';
+
+describe('api', () => {
```
api 에 테스트를 잘 작성하셨어요! 👏 

테스트를 작성하시면서 무엇을 확인하고 싶으셨나요?
fetch 함수가 올바르게 동작하는지? 요청에 대한 응답이 잘 오는지?

굳이 확인하지 않아도 되는 불필요한 테스트라는 생각이 들죠. 이를 통해 커버리지 100% 가 테스트의 품질을 보장해주진 않는다는 점을 알 수 있어요! 

물론 과제에선 테스트코드 작성 훈련을 위해 100% 커버리지를 요구하고 있지만요 :)

passwd10 commented on src/reducer.test.js:
```diff
@@ -0,0 +1,112 @@
+      selectedCategory: null,
+    };
+
+    it('has the initial state', () => {
```
reducer가 실제 하는역할은 초기상태를 반환하는 일이기 때문에 `returns initial state` 로 표현하면 좋을 것 같아요.

passwd10 commented on src/async-actions.test.js:
```diff
@@ -0,0 +1,29 @@
+    it('dispatches setRegions', async () => {
+      await loadRegions()(dispatch);
+
+      expect(dispatch).toBeCalledWith(setRegions(regions));
```
`redux-mock-store` 라이브러리를 사용하면 dispatch 시에 어떤 action 을 호출했는지 확인할 수 있어요

passwd10 reviewed (commented):
모든 게 올바르게 동작하는 것을 가정하는 Happy path에 대해서 먼저 테스트를 작성하시고 케이스를 하나씩 구체화해가면서 진행해보세요.

사용자에게 동작하는 어플리케이션을 보여주려면 무엇부터 테스트 해야할까요?

힌트를 드리자면 App에 대한 테스트부터 시작해보시면 좋을 것 같아요 :)

2scent commented:
이전 과제까지 reducer 코드부터 작성해서 이번 과제에서는 App부터 작성해보려고 했는데요. 막상 App부터 시작하려니 한 번에 여러 코드를 건드리게 돼서 커밋을 어떻게 나눠야 할지 모르겠더라고요. 이에 대해서 팁 좀 주실 수 있을까요?

2scent committed:
Refactor async-actions test

to use redux-mock-store

2scent committed:
Rename reducer test without state

2scent commented:
오늘은 일이 있어서 피드백 주신 것 말고는 작업을 거의 못했네요.

알려주신 `redux-mock-store` 사용해봤는데 이런 식으로 사용하면 되나요?

passwd10 commented on src/async-actions.test.js:
```diff
@@ -6,24 +9,28 @@ import { fetchRegions } from './services/api';
 
 jest.mock('./services/api');
 
+const mockStore = configureStore([thunk]);
```
mockStore 를 잘 사용하고 계시네요 👍

passwd10 reviewed (commented):
지금 당장 사용자에게 레스토랑 목록을 보여줘야한다면 무엇부터 시작하면 좋을까요?

- `RED` 사용자가 App 에서 레스토랑 목록을 볼 수 있는지 테스트를 먼저 작성할 것 같아요. 
- `GREEN` 작성한 테스트를 가장 빨리 통과시키려면 레스토랑 목록을 바로 App 컴포넌트에서 표현해주면 되겠죠. 
- `REFACTOR` 리팩토링이 필요하다고 생각되면 테스트가 깨지지 않는 범위내에서 리팩토링을 진행하는거에요.

첫번째 테스트에선 단순히 목록을 보여주기만 하면 되기 때문에 데이터를 fetching할 필요도 없고 Redux를 사용할 필요도 없어요.

만약 `실시간으로 레스토랑 목록을 보여줘야한다` 라는 테스트가 추가된다면 그때에 데이터 fetching 작업을 진행하면 되는거에요.

`RED`-`GREEN`-`REFACTOR` 싸이클을 반복하면서 최대한 작게 만들어서 빠르게 테스트를 통과시킬 수 있도록 해보세요 ㅎㅎ

2scent committed:
Add test for rendering regions

2scent committed:
Edit App for rendering regions

2scent committed:
Edit App test for rendering fetched regions

2scent committed:
Edit App for rendering fetched regions

2scent committed:
Refactor regions feature

App에서 하던 regions를 가져오고, 보여주던 기능을
RegionsContainer와 Regions로 분리

2scent committed:
Add Region

Region 추가하고, Regions에서 렌더하도록 수정

2scent committed:
Add feature for dispatching selectRegion

RegionsContainer에 selectRegion 관련 코드 및 테스트 추가
Region에서 onClick에 region을 넘겨주도록 코드 및 테스트 수정

2scent committed:
Add test for listening click event

2scent commented:
피드백 주신 내용 참고해서 `App` 컴포넌트를 시작으로 regions 관련된 코드를 작성해봤습니다.

비동기 작업이 들어가니깐 컴포넌트 테스트를 작성하는 게 어려워진 것 같은데, 잘하고 있는 걸까요?

passwd10 commented on src/App.test.jsx:
```diff
@@ -1,13 +1,33 @@
 import App from './App';
 
+jest.mock('react-redux');
+jest.mock('./async-actions');
```
어떤 의도로 async-actions 를 mocking 하려 하신건지 알 수 있을까요~?
jest.mock 을 하면 어떤일이 발생할까요?

2scent replied:
아래 피드백 주신 내용들 포함해서 답변드리자면, 지역 목록을 가져오는 기능을 `App`에 먼저 작성하고, `RegionsContainer`로 분리했는데 테스트 코드까지 신경을 못 썼던 것 같습니다. 지금은 수정했습니다.

passwd10 commented on src/App.test.jsx:
```diff
@@ -1,13 +1,33 @@
+  it('renders fetched regions', () => {
     const { container } = render(<App />);
 
+    expect(dispatch).toHaveBeenCalledWith(loadRegions());
```
해당 테스트가 렌더링을 확인하는 테스트 하위에 위치하는게 문맥상 맞지 않는 것 같아요!

loadRegion 이 dispatch 되었는지를 검증하는 테스트를 따로 분리해보시는건 어떤가요?

passwd10 commented on src/App.test.jsx:
```diff
@@ -1,13 +1,33 @@
+    jest.clearAllMocks();
+  });
+
+  it('renders fetched regions', () => {
```
렌더링하려는 regions가 `fetch` 된 상태인지 아닌지를 컴포넌트가 알 필요가 있을까요? ㅎㅎ 해당 컴포넌트의 관심사를 생각해보시면 좋을 것 같아요!

passwd10 commented on src/Regions.test.jsx:
```diff
@@ -1,21 +1,40 @@
   });
+
+  it('listens click event', () => {
+    const region = regions[2];
```
특별히 regions[2] 를 사용하신 이유가 있을까요~? 그냥 의도가 궁금해서 여쭤봐요!

2scent replied:
딱히 별 이유 없이 아무거나 쓴 겁니다.. 😅

passwd10 reviewed (commented):
쉽지 않으셨을텐데 피드백 드린 부분을 바로 적용해서 과제 진행하시는 모습 정말 훌륭해요 👍👍 

지금처럼 작고 빠르게 검증하는 싸이클을 반복하면서 점진적으로 케이스를 늘려가시면 될 것 같아요!

2scent committed:
Edit App test

regions가 렌더되는지만 확인하도록 수정

2scent committed:
Add test for rendering categories

2scent committed:
Edit App for rendering categories

2scent committed:
Edit App test for rendering fetched categories

2scent committed:
Edit App for rendering fetched categories

비동기 작업 때문에 test 코드도 waitFor()을 사용하도록 일부 수정

2scent committed:
Add loadCategories async action

2scent committed:
Refactor App to use redux

2scent committed:
Refactor categories feature

Category, Categoreis, CategoriesContainer 추가
App에서 CategoriesContainer 사용하도록 수정

2scent committed:
Edit Regions for rendering selected region

selectedRegion이 있다면 체크 마크를 표시하는지 확인하는 테스트 및 기능
추가

2scent committed:
Edit RegionsContainer for rendering selected region

2scent committed:
Edit Categories for rendering selected category

2scent committed:
Edit CategoriesContainer for rendering selected category

2scent committed:
Add RestaurantsContainer

우선 기본 식당 목록 렌더하도록 함

2scent committed:
Refactor fetchRestaurants

RORO 패턴 사용

2scent committed:
Add searchRestaurants action

2scent committed:
Edit to render searched restaurants

2scent commented:
피드백 주신 내용들 반영하고, 지역이랑 분류 목록 보여주고, 선택하는 것까지 구현했습니다.

현재 레스토랑 목록 가져오는 기능 구현 중인데, 테스트 코드 작성에서 어려움이 있는데요.

```javascript
  useEffect(() => {
    dispatch(searchRestaurants({ region, category }));
  }, [region, category]);
```

위와 같이 `RestaurantsContainer`에서 `region` 또는 `category`가 바뀔 때마다 레스토랑 목록을 검색하도록 하고 싶은데요. 이에 대한 테스트 코드를 어떻게 작성해야할지 감이 잘 안 잡히는데 어떤 식으로 할 수 있을까요?

passwd10 commented on src/Regions.test.jsx:
```diff
@@ -29,12 +32,16 @@ describe('Regions', () => {
+    const { getByRole } = renderRegions({ selectedRegion: initialRegion });
 
-    expect(handleClick).toHaveBeenCalledWith(region);
+    expect(getByRole('button', { name: `${initialRegion.name}(V)` })).toBeInTheDocument();
```
`container`와 `toHaveTextContent` 를 이용하면 더 직관적으로 표현할 수 있을 것 같은데 어떻게 생각하시나요 ㅎㅎ

passwd10 commented on src/RestaurantsContainer.jsx:
```diff
@@ -0,0 +1,27 @@
+
+  useEffect(() => {
+    dispatch(searchRestaurants({ region, category }));
+  }, []);
```
의존성 배열이 비어있네요! 적절한 상태를 넣어줘야 원하는 결과를 얻을 수 있을 것 같아요

2scent replied:
테스트를 작성하기 전이라 일부러 비워 놨습니다.

passwd10 reviewed (commented):
region 또는 category 가 바뀔 때에 dispatch가 잘 실행되는지를 확인하면 되지 않을까요?

2scent committed:
Edit restaurants features

RestaurantsContainer에서 View와 관련된 기능들은 Restaurants로 분리
selectedRegion, selectedCategory 없으면 레스토랑 검색 안 되도록 수정

2scent committed:
Refactor test queries

체크 표시를 확인할 때, container와 toHaveTextContent 사용

2scent commented:
피드백 주신 것 반영하고, `selectedRegion`이랑 `selectedCategory`가 바뀔 때마다 레스토랑 목록을 새로 가져오도록 기능은 구현했는데, 저것들이 바뀐다는 것을 테스트로 어떻게 작성해야할지 모르겠네요.

passwd10 commented:
> 피드백 주신 것 반영하고, `selectedRegion`이랑 `selectedCategory`가 바뀔 때마다 레스토랑 목록을 새로 가져오도록 기능은 구현했는데, 저것들이 바뀐다는 것을 테스트로 어떻게 작성해야할지 모르겠네요.

말씀하신 테스트는 e2e 테스트에 가까운 것 같아요.
region과 category가 바뀌려면 사용자가 어떤행동을 해야하는지 생각해보시면 좋을 것 같아요!

2scent committed:
Edit e2e test

api 응답값 변경으로 인한 e2e 테스트 수정

2scent commented:
region과 category가 바뀌려면, 그에 해당하는 버튼을 눌러야 하는데요. 이를 위해서는 App 컴포넌트에서 테스트를 작성해야 할 것 같은데요.  RestaurantsContainer의 useEffect에 region 또는 category의 값이 변경됐다는 것을 전해주기 어렵다는 같은 문제가 있고, 관심사도 아닌 것 같다는 생각이 드네요.

이런 경우는 e2e 테스트로만 확인할 수 있는 건가요?

passwd10 commented:
맞아요. 외부 컴포넌트에 의존해야하다는건 해당 컴포넌트의 관심사가 아닐 확률이 높아요. 각각의 컴포넌트는 독립적으로 격리해서 테스트해야해요. 실제 RestaurantsContainer 를 보면 store 에서 가져온 상태를 이용해서 액션을 dispatch 를 해주는것뿐이지, `상태가 바뀌는 것`은 RestaurantsContainer에서 컨트롤하고 있지않기 때문에 테스트가 안될거에요 ㅎㅎ

저는 해당 상황처럼 컴포넌트간의 상호작용이 필요한 경우(일반적으로 사용자의 flow 검증) E2E 테스트를 활용하는 편이에요!
